wb_modbus_device_editor/main.py

```diff
@@ -253,129 +267,167 @@ def widgets_hide_by_condition(self):
+                failed = True
             else:
-                self.ui.write_log("Не прочитал ни одного параметра")
-
-            self.widgets_hide_by_condition()
+                self.ui.set_value(id, value, scale=param.get("scale"))
+                self.ui.widget_enable(id)
+        if failed:
+            self.ui.write_log(
+                "Не удалось прочитать некоторые параметры. Возможно их нет в этой версии прошивки устройства. Такие параметры будут недоступны для редактирования."
+            )
         else:
-            self.ui.write_log("Не смог открыть порт {}".format(mb_params["port"]))
-        client.disconnect()
+            self.widgets_hide_by_condition()
+            self.ui.write_log("Чтение параметров завершено")
+        self.client.disconnect()
 
-    def prepare_address(self, address):
-        if isinstance(address, str):
-            return int(address, 16)
-        else:
-            return int(address)
+        with self.io_lock:
+            self.io_running = False
 
-    def prepare_reg_type(self, reg_type):
-        if reg_type is None:
-            return "holding"
-        else:
-            return reg_type
+    def read_params_from_modbus_errback(self, error):
+        self.ui.write_log(f"Ошибка во время чтения параметов: {error}")
+        self.client.disconnect()
+
+        with self.io_lock:
+            self.io_running = False
 
     def read_params_from_modbus(self, client, slave_id, params):
-        cnt = 0
-        if len(params) > 0:
-            for i, _ in enumerate(params):
-                param = params[i]
-
-                # адреса пишут то в HEX то в DEC, надо определять и преобразовывать в DEC
-                address = self.prepare_address(param["address"])
-
-                # бывает, что тип регистра не пишет, надо присвоить значение по умолчанию
-                reg_type = self.prepare_reg_type(param.get("reg_type"))
-
-                if reg_type == "holding":
-                    try:
-                        value = client.read_holding(slave_id=slave_id, reg_address=address)
-
-                        self.ui.set_value(param["id"], value, scale=param.get("scale"))
-                        self.ui.widget_enable(param["id"])
-                        cnt += 1
-                    except Exception as e:
-                        if "ExceptionResponse" in "%s" % e:
-                            self.ui.widget_disable(param["id"])
-                            self.ui.write_log("Регистр {} не читается.".format(address))
-                        if "ModbusIOException" in "%s" % e:
-                            self.ui.write_log("Нет связи с устройством.")
-                            break
-            return cnt
+        result = []
+
+        for id, param in params.items():
+            # адреса пишут то в HEX то в DEC, надо определять и преобразовывать в DEC
+            address = (
+                int(param["address"], 16) if isinstance(param["address"], str) else int(param["address"])
```

Comment:
```suggestion
                int(param["address"], 0) if isinstance(param["address"], str) else int(param["address"])
```

или у нас в string только вариант с `0x`?

Reply:
поправлено